File: src/AdventOfCode2015/Puzzles/Day08.cs
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

namespace MartinCostello.AdventOfCode2015.Puzzles
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using System.Text;

    /// <summary>
    /// A class representing the puzzle for <c>http://adventofcode.com/day/8</c>. This class cannot be inherited.
    /// </summary>
    internal sealed class Day08 : IPuzzle
    {
        /// <summary>
        /// Gets the value for the first solution.
        /// </summary>
        internal int FirstSolution { get; private set; }

        /// <summary>
        /// Gets the value for the second solution.
        /// </summary>
        internal int SecondSolution { get; private set; }

        /// <inheritdoc />
        public int Solve(string[] args)
        {
            if (args.Length != 1)
            {
                Console.Error.WriteLine("No input file path specified.");
                return -1;
            }

            if (!File.Exists(args[0]))
            {
                Console.Error.WriteLine("The input file path specified cannot be found.");
                return -1;
            }

            string[] input = File.ReadAllLines(args[0]);

            int countForCode = input.Sum((p) => p.Length);
            int countInMemory = GetLiteralCharacterCount(input);
            int countEncoded = GetEncodedCharacterCount(input);

            Console.WriteLine(
                "The number of characters of code for string literals minus the number of characters in memory for the values of the strings is {0:N0}.",
                FirstSolution = countForCode - countInMemory);

            Console.WriteLine(
                "The total number of characters to represent the newly encoded strings minus the number of characters of code in each original string literal is {0:N0}.",
                SecondSolution = countEncoded - countForCode);

            return 0;
        }

        /// <summary>
        /// Returns the number of characters in the specified collection of <see cref="string"/> if literals are encoded.
        /// </summary>
        /// <param name="collection">The values to get the encoded number of characters from.</param>
        /// <returns>The number of characters in <paramref name="collection"/> when encoded.</returns>
        internal static int GetEncodedCharacterCount(IEnumerable<string> collection) => collection.Sum(GetEncodedCharacterCount);

        /// <summary>
        /// Returns the number of characters in the specified <see cref="string"/> if literals are encoded.
        /// </summary>
        /// <param name="value">The value to get the encoded number of characters from.</param>
        /// <returns>The number of characters in <paramref name="value"/> when encoded.</returns>
        internal static int GetEncodedCharacterCount(string value)
        {
            StringBuilder builder = new StringBuilder("\"");

            for (int i = 0; i < value.Length; i++)
            {
                char current = value[i];

                switch (current)
                {
                    case '\"':
                    case '\\':
                    case '\'':
                        builder.Append("\\");
                        break;

                    default:
                        break;
                }

                builder.Append(current);
            }

            builder.Append("\"");

            return builder.Length;
        }

        /// <summary>
        /// Returns the number of literal characters in the specified collection of <see cref="string"/>.
        /// </summary>
        /// <param name="collection">The values to get the number of literal characters from.</param>
        /// <returns>The number of literal characters in <paramref name="collection"/>.</returns>
        internal static int GetLiteralCharacterCount(IEnumerable<string> collection) => collection.Sum(GetLiteralCharacterCount);

        /// <summary>
        /// Returns the number of literal characters in the specified <see cref="string"/>.
        /// </summary>
        /// <param name="value">The value to get the number of literal characters from.</param>
        /// <returns>The number of literal characters in <paramref name="value"/>.</returns>
        internal static int GetLiteralCharacterCount(string value)
        {
            int count = 0;

            // Remove quotes if present as first/last characters
            bool removeFirstQuote = value.Length > 0 && value.First() == '\"';
            bool removeLastQuote = value.Length > 1 && value.Last() == '\"';

            if (removeFirstQuote)
            {
                value = value.Substring(1);
            }

            if (removeLastQuote)
            {
                value = value.Substring(0, value.Length - 1);
            }

            if (value.Length > 0)
            {
                Queue<char> characters = new Queue<char>(value);

                while (characters.Count > 0)
                {
                    char current = characters.Dequeue();

                    if (characters.Count > 0)
                    {
                        switch (current)
                        {
                            case '\\':
                                char next = characters.Peek();

                                if (next == '\"' || next == '\'' || next == '\\')
                                {
                                    characters.Dequeue();
                                }
                                else if (next == 'x' && characters.Count > 2)
                                {
                                    characters.Dequeue();
                                    characters.Dequeue();
                                    characters.Dequeue();
                                }

                                break;

                            default:
                                break;
                        }
                    }

                    count++;
                }
            }

            return count;
        }
    }
}

File: src/AdventOfCode.Site/scripts/ts/puzzles/2015/Day08.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { from } from 'linq-to-typescript';
import { Puzzle2015 } from './Puzzle2015';

export class Day08 extends Puzzle2015 {
    firstSolution: number;
    secondSolution: number;

    override get name() {
        return 'Matchsticks';
    }

    override get day() {
        return 8;
    }

    static getEncodedCharacterCount(...collection: string[]) {
        let count = 0;

        for (const value of collection) {
            const encoded: Array<string> = ['"'];

            for (let i = 0; i < value.length; i++) {
                let current = value[i];

                switch (current) {
                    case '"':
                    case '\\':
                    case "'":
                        encoded.push('\\');
                        break;

                    default:
                        break;
                }

                encoded.push(current);
            }

            encoded.push('"');

            count += encoded.length;
        }

        return count;
    }

    static getLiteralCharacterCount(...collection: string[]) {
        let result = 0;

        for (let value of collection) {
            let count = 0;

            // Remove quotes if present as first/last characters
            let removeFirstQuote = value.length > 0 && value[0] === '"';
            let removeLastQuote = value.length > 1 && value[value.length - 1] === '"';

            if (removeFirstQuote) {
                value = value.slice(1);
            }

            if (removeLastQuote) {
                value = value.slice(0, -1);
            }

            if (value.length > 0) {
                const characters: Array<string> = [];

                for (let i = 0; i < value.length; i++) {
                    characters.push(value[i]);
                }

                while (characters.length > 0) {
                    let current = characters.shift();

                    if (characters.length > 0) {
                        switch (current) {
                            case '\\':
                                const next = characters[0];

                                if (next === '"' || next === "'" || next === '\\') {
                                    characters.shift();
                                } else if (next === 'x' && characters.length > 2) {
                                    characters.shift();
                                    characters.shift();
                                    characters.shift();
                                }

                                break;

                            default:
                                break;
                        }
                    }

                    count++;
                }
            }

            result += count;
        }

        return result;
    }

    override solveCore(_: string[]) {
        const input = this.readResourceAsLines();

        const countForCode = from(input).sum((p: string) => p.length);
        const countInMemory = Day08.getLiteralCharacterCount(...input);
        const countEncoded = Day08.getEncodedCharacterCount(...input);

        this.firstSolution = countForCode - countInMemory;
        this.secondSolution = countEncoded - countForCode;

        console.info(
            `The number of characters of code for string literals minus the number of characters in memory for the values of the strings is ${this.firstSolution}.`
        );
        console.info(
            `The total number of characters to represent the newly encoded strings minus the number of characters of code in each original string literal is ${this.secondSolution}.`
        );

        return this.createResult([this.firstSolution, this.secondSolution]);
    }
}
